sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala

```diff
@@ -211,4 +215,80 @@ trait ExpressionEvalHelper {
       plan(inputRow)).get(0, expression.dataType)
     assert(checkResult(actual, expected))
   }
+
+  def checkConsistency(dt: DataType, clazz: Class[_]): Unit = {
+    val ctor = clazz.getDeclaredConstructor(classOf[Expression])
+    forAll (randomGen(dt)) { (l: Literal) =>
+      val expr = ctor.newInstance(l).asInstanceOf[Expression]
+      cmpInterpretWithCodegen(EmptyRow, expr)
+    }
+  }
+
+  def checkConsistency(dt1: DataType, dt2: DataType, clazz: Class[_]): Unit = {
+    val ctor = clazz.getDeclaredConstructor(classOf[Expression], classOf[Expression])
+    forAll (
+        randomGen(dt1),
+        randomGen(dt2)
+    ) { (l1: Literal, l2: Literal) =>
+      val expr = ctor.newInstance(l1, l2).asInstanceOf[Expression]
+      cmpInterpretWithCodegen(EmptyRow, expr)
+    }
+  }
+
+  def checkConsistency(dt1: DataType, dt2: DataType, dt3: DataType,
+      clazz: Class[_]): Unit = {
+    val ctor = clazz.getDeclaredConstructor(
+      classOf[Expression], classOf[Expression], classOf[Expression])
+    forAll (
+        randomGen(dt1),
+        randomGen(dt2),
+        randomGen(dt3)
+    ) { (l1: Literal, l2: Literal, l3: Literal) =>
+      val expr = ctor.newInstance(l1, l2, l3).asInstanceOf[Expression]
+      cmpInterpretWithCodegen(EmptyRow, expr)
+    }
+  }
+
+  def checkSeqConsistency(dt: DataType, clazz: Class[_], leastNumOfElements: Int = 0): Unit = {
+    val ctor = clazz.getDeclaredConstructor(classOf[Seq[Expression]])
+    forAll (Gen.listOf(randomGen(dt))) { (literals: Seq[Literal]) =>
+      whenever(literals.size >= leastNumOfElements) {
+        val expr = ctor.newInstance(literals).asInstanceOf[Expression]
+        cmpInterpretWithCodegen(EmptyRow, expr)
+      }
+    }
+  }
+
+  private def cmpInterpretWithCodegen(inputRow: InternalRow, expr: Expression): Unit = {
+    val interpret = try evaluate(expr, inputRow) catch {
+      case e: Exception => fail(s"Exception evaluating $expr", e)
+    }
+
+    val plan = generateProject(
+      GenerateMutableProjection.generate(Alias(expr, s"Optimized($expr)")() :: Nil)(),
+      expr)
+    val codegen = plan(inputRow).get(0, expr.dataType)
+
+    if (!checkResultRegardingNaN(interpret, codegen)) {
```

Comment:
`checkResultRegardingNaN` is slightly confusing.  Maybe we could just call this `compareResults`?